components/tikv_util/src/quota_limiter.rs

```diff
@@ -0,0 +1,176 @@
+#[derive(Debug)]
+pub struct QuotaLimiter {
+    cputime_limiter: Limiter,
+    write_bandwidth_limiter: Limiter,
+    read_bandwidth_limiter: Limiter,
+}
+
+impl Default for QuotaLimiter {
+    fn default() -> Self {
+        Self {
+            cputime_limiter: Limiter::new(f64::INFINITY),
+            write_bandwidth_limiter: Limiter::new(f64::INFINITY),
+            read_bandwidth_limiter: Limiter::new(f64::INFINITY),
+        }
+    }
+}
+
+impl QuotaLimiter {
+    // 1000 millicpu equals to 1vCPU, 0 means unlimited
+    pub fn new(
+        cpu_quota: usize,
+        write_bandwidth: ReadableSize,
+        read_bandwidth: ReadableSize,
+    ) -> Self {
+        let cputime_limiter = if cpu_quota == 0 {
+            Limiter::new(f64::INFINITY)
+        } else {
+            Limiter::new(cpu_quota as f64 * CPU_TIME_FACTOR * 1000_f64)
+        };
+
+        let write_bandwidth_limiter = if write_bandwidth.0 == 0 {
+            Limiter::new(f64::INFINITY)
+        } else {
+            Limiter::new(write_bandwidth.0 as f64)
+        };
+
+        let read_bandwidth_limiter = if read_bandwidth.0 == 0 {
+            Limiter::new(f64::INFINITY)
+        } else {
+            Limiter::new(read_bandwidth.0 as f64)
+        };
+
+        Self {
+            cputime_limiter,
+            write_bandwidth_limiter,
+            read_bandwidth_limiter,
+        }
+    }
+
+    // record info after write requests finished and return the suggested delay duration.
+    pub fn consume_write(&self, bytes: usize, cpu_time: Duration) -> Duration {
+        let cpu_dur = self
+            .cputime_limiter
+            .consume_duration(cpu_time.as_micros() as usize);
+
+        let bw_dur = if bytes > 0 {
+            self.write_bandwidth_limiter.consume_duration(bytes)
+        } else {
+            Duration::ZERO
+        };
+
+        let max_dur = std::cmp::max(cpu_dur, bw_dur);
+        let should_delay = if max_dur > cpu_time {
+            max_dur - cpu_time
+        } else {
+            Duration::ZERO
+        };
+        std::cmp::min(MAX_QUOTA_DELAY, should_delay)
+    }
+
+    // record info after read requests finished and return the suggested delay duration.
+    pub fn consume_read(&self, bytes: usize, cpu_time: Duration) -> Duration {
+        let cpu_dur = self
+            .cputime_limiter
+            .consume_duration(cpu_time.as_micros() as usize);
+
+        let bw_dur = if bytes > 0 {
+            self.read_bandwidth_limiter.consume_duration(bytes)
+        } else {
+            Duration::ZERO
+        };
+
+        let max_dur = std::cmp::max(cpu_dur, bw_dur);
+        let should_delay = if max_dur > cpu_time {
+            max_dur - cpu_time
+        } else {
+            Duration::ZERO
+        };
+        std::cmp::min(MAX_QUOTA_DELAY, should_delay)
+    }
+
+    // If `cputime_limiter` is set to INFINITY, use `CLOCK_MONOTONIC_COARSE` to save cost.
+    pub fn get_now_timer(&self) -> Box<dyn Timer> {
```

Comment:
Could returning an `Option` be enough?

Reply:
fixed in https://github.com/tikv/tikv/pull/12151#discussion_r827606925